base/threadingconstructs.jl

```diff
@@ -3,24 +3,59 @@
 export threadid, nthreads, @threads, @spawn
 
 """
-    Threads.threadid()
+    Threads.threadid() -> Int
 
-Get the ID number of the current thread of execution. The master thread has ID `1`.
+Get the ID number of the current thread of execution within the current
+threadpool. The master thread has ID `1`.
 """
-threadid() = Int(ccall(:jl_threadid, Int16, ())+1)
+threadid() = Int(ccall(:jl_relative_threadid, Int16, ())+1)
 
-# Inclusive upper bound on threadid()
 """
-    Threads.nthreads()
+    Threads.threadid() -> Int
 
-Get the number of threads available to the Julia process. This is the inclusive upper bound
-on [`threadid()`](@ref).
+Get the ID number of the current thread of execution within the Julia session.
+The master thread has ID `1`.
+"""
+rawthreadid() = Int(ccall(:jl_threadid, Int16, ())+1)
+
+"""
+    Threads.nthreads(tp::Int=Threads.threadpoolid()) -> Int
+
+Get the number of threads available in the specified threadpool. This is the
+inclusive upper bound on [`threadid()`](@ref).
 
 See also: `BLAS.get_num_threads` and `BLAS.set_num_threads` in the
 [`LinearAlgebra`](@ref man-linalg) standard library, and `nprocs()` in the
 [`Distributed`](@ref man-distributed) standard library.
 """
-nthreads() = Int(unsafe_load(cglobal(:jl_n_threads, Cint)))
+nthreads(tp::Int=Threads.threadpoolid()) =
+    ccall(:jl_num_threads, Cint, (Cint,), tp-1)
+
+"""
+    Threads.spawn_threadpool(n::Int) -> Int
+
+Spawns a new threadpool of size `n`, and returns the threadpool ID.
+"""
+function spawn_threadpool(n::Int)
+    wq = [Base.StickyWorkqueue() for _ in 1:n]
+    push!(Base.AllWorkqueues, wq)
```

Comment:
This `push!` is racy since it could resize the internal buffer of `AllWorkqueues`. We'd need a lock on `AllWorkqueues` or use a proper concurrent data structure.

Reply:
Also, other things in `jl_start_threads_` (called from here) like `jl_threadpools++` are also racy. I guess we also need to update `jl_all_tls_states` here which is tricky (since GC can start anytime).

Actually, since the nthreads for this new pool needs to match in the Julia and C side, protecting `AllWorkqueues` is not enough. Maybe it's easier to just use a lock here and making sure to acquire it before every time `jl_start_threads_dedicated` is called. But synchronization of `jl_all_tls_states` still needs something else...

Reply:
I think what we can do on the runtime side is to only (atomically) increment `jl_threadpools` once all the runtime and Julia pieces are fully setup, and have all arrays be fully sized (to the max number of threads/threadpools we'll allow, which can be a build-time parameter) at init time. I also agree that `jl_start_threads_dedicated` can take a lock to ensure serialization.

Reply:
I think the fun(?) part is `jl_all_tls_states`. I'd imagine something like this (in C)

```julia
new = copy(jl_all_tls_states)
new_ptls = [create_ptls() for _ in 1:n]
append!(new, new_ptls)

# No need for CAS since `jl_start_threads_dedicated` (or its caller)
# acquires the lock
@atomic jl_all_tls_states = new

# Here, we still can't free `old` because there can be
# `jl_gc_wait_for_the_world` etc.

GC.gc(false)
# Once we run GC and stooped all existing thread, we can be sure there
# are no stale accesses to `jl_all_tls_states`
free(old)
```

...or maybe just leak old `jl_all_tls_states` to simply things.

Alternatively, we can use linked list (of blocks) for `jl_all_tls_states`.

Reply:
@tkf now that threadpool spawning is done only at startup, would you mind re-reviewing? I'm not sure what kinds of atomics `t->threadpool = xyz` needs (in the runtime).